store/tikv/batch_coprocessor.go

```diff
@@ -0,0 +1,422 @@
+}
+
+func (b *batchCopIterator) recvFromRespCh(ctx context.Context) (resp *batchCopResponse, ok bool, exit bool) {
+	select {
+	case resp, ok = <-b.respChan:
+	case <-b.finishCh:
+		exit = true
+	case <-ctx.Done():
+		// We select the ctx.Done() in the thread of `Next` instead of in the worker to avoid the cost of `WithCancel`.
+		if atomic.CompareAndSwapUint32(&b.closed, 0, 1) {
+			close(b.finishCh)
+		}
+		exit = true
+	}
+	return
+}
+
+// Close releases the resource.
+func (b *batchCopIterator) Close() error {
+	if atomic.CompareAndSwapUint32(&b.closed, 0, 1) {
+		close(b.finishCh)
+	}
+	b.wg.Wait()
+	return nil
+}
+
+func (b *batchCopIterator) handleTask(ctx context.Context, bo *Backoffer, task *batchCopTask) {
+	logutil.BgLogger().Debug("handle batch task")
+	tasks := []*batchCopTask{task}
+	idx := 0
+	for idx < len(tasks) {
+		ret, err := b.handleTaskOnce(ctx, bo, task)
+		if err != nil {
+			resp := &batchCopResponse{err: errors.Trace(err)}
+			b.sendToRespCh(resp)
+		} else {
+			tasks = append(tasks, ret...)
+		}
+		idx++
+	}
+	b.wg.Done()
+}
+
+// Merge all ranges and request again.
+func (b *batchCopIterator) retryBatchCopTask(ctx context.Context, bo *Backoffer, batchTask *batchCopTask) ([]*batchCopTask, error) {
+	ranges := &copRanges{}
+	for _, taskCtx := range batchTask.copTasks {
+		taskCtx.task.ranges.do(func(ran *kv.KeyRange) {
+			ranges.mid = append(ranges.mid, *ran)
+		})
+	}
+	return buildBatchCopTasks(bo, b.RegionCache, ranges, b.req)
+}
+
+func (b *batchCopIterator) handleTaskOnce(ctx context.Context, bo *Backoffer, task *batchCopTask) ([]*batchCopTask, error) {
+	logutil.BgLogger().Debug("handle batch task once")
+	sender := NewRegionBatchRequestSender(b.store.regionCache, b.store.client)
+	var regionInfos []*coprocessor.RegionInfo
+	for _, task := range task.copTasks {
+		regionInfos = append(regionInfos, &coprocessor.RegionInfo{
+			RegionId: task.task.region.id,
+			RegionEpoch: &metapb.RegionEpoch{
+				ConfVer: task.task.region.confVer,
+				Version: task.task.region.ver,
+			},
+			Ranges: task.task.ranges.toPBRanges(),
+		})
+	}
+
+	copReq := coprocessor.BatchRequest{
+		Tp:        b.req.Tp,
+		StartTs:   b.req.StartTs,
+		Data:      b.req.Data,
+		SchemaVer: b.req.SchemaVar,
+		Regions:   regionInfos,
+	}
+
+	req := tikvrpc.NewReplicaReadRequest(task.cmdType, &copReq, b.req.ReplicaRead, b.replicaReadSeed, kvrpcpb.Context{
+		IsolationLevel: pbIsolationLevel(b.req.IsolationLevel),
+		Priority:       kvPriorityToCommandPri(b.req.Priority),
+		NotFillCache:   b.req.NotFillCache,
+		HandleTime:     true,
+		ScanDetail:     true,
+	})
+	req.StoreTp = kv.TiFlash
+
+	logutil.BgLogger().Debug("send batch request to ", zap.String("req info", req.String()), zap.Int("cop task len", len(task.copTasks)))
+	resp, retry, err := sender.sendReqToAddr(bo, task.copTasks, req, ReadTimeoutMedium)
+	// If there are store errors, we should retry for all regions.
+	if retry {
+		return b.retryBatchCopTask(ctx, bo, task)
```

Comment:
Is this function means that spilt one `batchCopTask` to many `batchCopTask` and try again?

Reply:
yes